Adding Environment Yaml file README - first draft

dncliss opened the pull request:
## Checklist before submitting

- [ x] Did you read the [contributor guide](https://github.com/open-ce/open-ce/blob/master/CONTRIBUTING.md)?
- [x ] Did you update any affected [documentation](https://github.com/open-ce/open-ce/blob/master/doc/)?
- [n/a ] Did you write any [tests](https://github.com/open-ce/open-ce/blob/master/tests/) to validate this change?  

## Description

Adds a README file describing (at a high level) the YAML files that we use in Open-CE.

Intended for Issue https://github.com/open-ce/open-ce/issues/51

dncliss commented:
With a file like this, I could go on and on making tweaks indefinitely.... so I needed to just go ahead and put something forward as the first version.

However, this first push version has become partly obsolete immediately; it includes a description of `channels` in the environment files, but a PR was just put in earlier today which removed all of the `channels` references from all of the environment files, making that description no longer necessary.  So, I'll update this PR again here, removing that part of the environment file description.

bethune-bryant commented on doc/README.yaml.md:
````diff
@@ -0,0 +1,230 @@
+[xgboost-env.yaml](https://github.com/open-ce/open-ce/blob/master/envs/xgboost-env.yaml):
+```
+packages:
+  - feedstock : nccl              #['cuda' in build_type]
````
This should now be `==` not `in`.
https://github.com/open-ce/open-ce/blob/master/envs/xgboost-env.yaml
```suggestion
  - feedstock : nccl              #[build_type == 'cuda']
```

bethune-bryant commented:
> 
> However, this first push version has become partly obsolete immediately; it includes a description of `channels` in the environment files, but a PR was just put in earlier today which removed all of the `channels` references from all of the environment files, making that description no longer necessary.

Channels can still be specified. We just removed references to the PowerAI conda channel. The schema for the Open-CE env files is defined here:

https://github.com/open-ce/open-ce/blob/d0a2ce04e98015fbe7d0394cfe26a803241963ea/open-ce/env_config.py#L20..L45

dncliss commented:
> > However, this first push version has become partly obsolete immediately; it includes a description of `channels` in the environment files, but a PR was just put in earlier today which removed all of the `channels` references from all of the environment files, making that description no longer necessary.
> 
> Channels can still be specified. We just removed references to the PowerAI conda channel. The schema for the Open-CE env files is defined here:
> 
> https://github.com/open-ce/open-ce/blob/d0a2ce04e98015fbe7d0394cfe26a803241963ea/open-ce/env_config.py#L21-L45

I agree that channels can still be specified, but I was taking this more to be documenting what we are actually using in the open-ce env files (omitting discussion of features not being used, since this isn't a full yaml specification). After the earlier PR today it looks like we don't have any env.yaml files that use the channels
specifier any more, so I removed it on purpose. (Our current env yaml files are pretty simplistic; the meta.yaml feedstock files are where the real action is, of course.)
Are you thinking there's a strong chance that channels is something people will want to add if customizing their own versions of our env files?  i.e. warranting inclusion in the README?

dncliss commented:
Regarding Bryant's `channels` comment above:
Per an offline conversation today (Oct 20), we decided that the README should still include a description of `channels` even though we have removed use, as people may want to use this; so I will add back the description of that tag.
I won't put it back quite exactly as before (it's no longer an actually-used example), but I think it should be pretty close.

bnemanich commented on doc/README.yaml.md:
```diff
@@ -0,0 +1,240 @@
+guides available elsewhere online.
+
+
+# build_env environment YAML file
```
The following section should be in a separate file. I'm not positive if we need it, or if we should just point to the conda documentation. But it might be nice to have still.

bnemanich replied:
Open-CE Environment File

bnemanich commented on doc/README.yaml.md:
```diff
@@ -0,0 +1,240 @@
+
+# README for Open-CE build YAML files
```
These should be titled "Open-CE Environment files". YAML can be left out of the title.

dncliss replied:
Done (here and universally).

bnemanich commented on doc/README.yaml.md:
```diff
@@ -0,0 +1,240 @@
+
+
+This README documents the YAML files used by the build scripts
+[build_env.py](https://github.com/open-ce/open-ce/blob/master/doc/README.build_env.md)
```
Links should be relative instead of picking a particular branch.

dncliss replied:
Done (here and universally in the file), for all cases within the open-ce repo itself.  (Exception is URLs in the file that point to a separate feedstock repository not within open-ce, which are general and not branch-specific.)

bnemanich commented on doc/README.yaml.md:
```diff
@@ -0,0 +1,240 @@
+and [build_feedstock.py](https://github.com/open-ce/open-ce/blob/master/doc/README.build_feedstock.md).
+Please refer to the respective README files for each of these two scripts.
+
+Most likely you will want to simply use the YAML files provided within the
```
YAML with Open-CE Environment

bnemanich commented on doc/README.yaml.md:
```diff
@@ -0,0 +1,240 @@
+do this, you will likely want to start by copying one of the existing files and
+making local edits as desired.
+
+NOTE: This file only documents information about the YAML files as used within
```
This NOTE is not needed.

dncliss replied:
Removed the NOTE. Thought it might be helpful to point to external official YAML documentation.... but then again if someone is customizing these files they probably know how to find such info if they need.

bnemanich commented on doc/README.yaml.md:
```diff
@@ -0,0 +1,240 @@
+file information is used from the yaml files of each respective feedstock repository.
+
+
+# build_feedstock YAML file (meta.yaml)
```
The following section should be in a separate file. I'm not positive if we need it, or if we should just point to the conda documentation. But it might be nice to have still.

bnemanich replied:
What we need documentation for is the build-config.yaml file that may appear in some feedstocks.

dncliss replied:
OK - this is definitely the biggest change, because it basically throws away the bottom 2/3 of the file, if we simply don't want to document this here.

Will anyone want this anyway, even if in a separate file?  We can leave it out (less to maintain) and nobody might care -- and if we decided to add it, the content should be retrievable from the file change history.

The one change *NOT* done here yet is the last part requesting documentation for feedstock build-config.yaml files, which is new content, so I need to write something up first.

bnemanich commented on doc/README.yaml.md:
```diff
@@ -0,0 +1,240 @@
+of its dependencies, each of which is found in a separate "feedstock" repository
+within the broader open-ce project.
+Default YAML files are provided for each of several environments in the open-ce
+repository, within https://github.com/open-ce/open-ce/tree/master/envs
```
Links should be relative.

bnemanich commented on doc/README.yaml.md:
```diff
@@ -0,0 +1,78 @@
+making local edits as desired.
+
+
+# Open-CE Environment File
```
There are several more possible fields in an Open-CE Environment file that aren't listed here. You can see a full list of them here: https://github.com/open-ce/open-ce/blob/master/open-ce/env_config.py#L32. The currently supported fields are:

packages:
    feedstock:
    recipes:
    channels:
    git_tag:
imported_envs:
git_tag_for_env:
external_dependencies:
channels:

jayfurmanek replied:
I agree, let's put this list at the top with quick explanations, then we can go into detail on each below like you already do with a few of them.

dncliss replied:
Done for all with exception of `external_dependencies` -- which I'm not seeing cases of how we use this anywhere within any builds -- only for a specific use within the open-ce/tests.  (So, am not sure how to reflect this one in a useful example).

bnemanich replied:
`external_dependencies` are the names of conda packages that will not be built, but will be added to the conda environment files that are generated by calling `build_env`.

jayfurmanek commented on doc/README.yaml.md:
```diff
@@ -0,0 +1,78 @@
+This is straightforward; it simply imports the contents of the listed environment
+files. So if you are building the horovod environment, you will get a build of
+both pytorch and tensorflow as well because of the specified imported
+environments listed in this section.
```
Let's refer to this as "nested" open-ce environment files.

jayfurmanek commented on doc/README.yaml.md:
```diff
@@ -0,0 +1,78 @@
+file information is used from the configuration files of each respective feedstock
+repository.
+
+
```
Let's add further descriptions here for the remaining available options from the top.

npanpaliya commented on doc/README.yaml.md:
```diff
@@ -0,0 +1,170 @@
+of its dependencies, each of which is found in a separate "feedstock" repository
+within the broader open-ce project.
+Default environment files are provided for each of several environments in the open-ce
+repository, within [https://github.com/open-ce/open-ce/envs](../envs)
```
https://github.com/open-ce/open-ce/envs -> This link isn't right. It should be https://github.com/open-ce/open-ce/tree/master/envs.

npanpaliya replied:
I just viewed this README from your branch. The links are correct but pointing to your branch. I think they will be updated once this is merged to master branch.

bnemanich commented:
Can you add some sort of header before the description of each field? Right now it is hard to find a description of what a particular field does without reading the entire file.
